Use percentages for relative heights of panels

sumwale opened the pull request:
* To avoid cutting off panels altogether use percentages for the various panels instead of fixed heights for all but process panel.
* By default the process panel occupies 10% more space than other non-zero ones.
* use a vector to track sections and their heights
  - use an ordered vector to track the geometry of each section (height for now) instead of using separate variables for each section
  - convert the height percentages into lengths before applying as constraints
  - catch terminal resize event and recalculate the lengths as per percentages
  - apply the constraints that non-process sections should have even number of rows, each section should be at least size 2, process section should have at least size 4; this is after accounting for all other section heights
  - pass a flag to check if process table had to "borrow" from others to keep its minimum size and abort size change due to e/m keys if that is the case

sumwale commented:
See discussion in the previous PR #67

sumwale commented:
@bvaisvil Can you review this PR and the other open one (#77)?

bvaisvil commented on src/main.rs:
```diff
@@ -178,16 +180,45 @@ fn start_zenith(
-            db,
-        );
+
+        let mut geometry: Vec<(Section, f64)> = Vec::new();
```
Could you refactor this section into a function? If possible, I'd also like the parameter checking to be done in main before `start_zenith` is called.

bvaisvil commented on src/main.rs:
```diff
@@ -178,16 +180,45 @@ fn start_zenith(
+        // sum of minimum percentages should not exceed 100%
+        let sum_heights = sum_section_heights(&geometry);
+        if sum_heights > 100.1 {
+            panic!(
```
Can you change the `panic!` for incorrect user specified parameters to use `println` and `exit`

sumwale replied:
Shouldn't it be fine here since panic_hook has already been setup. Or you want it to be so after refactor so that it can be invoked from main?

bvaisvil replied:
I'm not sure it's relevant anymore but I just thought it was bad form to present a panic to the user when what we really want to do is let them know the parameters they entered are in conflict.

bvaisvil reviewed (changes requested):
Thanks for making the changes, this looks good! There are a few more change I've noted, it would be great if you could update your PR with them. Thanks!

Cargo clippy has some recommended changes, you only have to worry about the ones added by your code. For example:
```
warning: writing `&Vec<_>` instead of `&[_]` involves one more reference and cannot be used with non-Vec-based slices.
 --> src/render.rs:93:38
93 | pub fn sum_section_heights(geometry: &Vec<(Section, f64)>) -> f64 {
   |                                      ^^^^^^^^^^^^^^^^^^^^ help: change this to: `&[(Section, f64)]`
   |
   = note: `#[warn(clippy::ptr_arg)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg
```
you can find the rest of them here: https://github.com/bvaisvil/zenith/pull/76/checks?check_run_id=1226735957

sumwale commented:
Pushed changes as per review. Also fixed a few existing clippy warnings. There are still three more about "this function has too many arguments". Added a macro "exit_with_message" to restore_terminal before exit consistently (also changed one exit in start_zenith to use the macro).

bvaisvil commented on src/main.rs:
```diff
@@ -107,6 +109,66 @@ fn use_db_history(db_path: &str, tick_rate: u64) -> Option<bool> {
     }
 }
 
+macro_rules! exit_with_message {
```
These changes look good!

bvaisvil commented:
Looks great, I'll merge right after your latest commit passes the checks. Thanks for the hard work!
